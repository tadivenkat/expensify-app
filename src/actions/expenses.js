import database from '../firebase/firebase';

export const addExpenseAction = (expense) => {
    return {
        type: 'ADD_EXPENSE',
        payload: expense
    };
};

export const startAddExpenseAction = (expenseData = {}) => {
    const {description = '', amount = 0, note = '', createdAt = 0} = expenseData;
    const expense = {description, amount, note, createdAt};
    return (dispatch) => {
        return database.ref('expenses').push(expense).then((data) => {
            dispatch(addExpenseAction({
                id: data.key,
                ...expense
            }));
        }).catch((error) => {
            console.log("Error adding expense", error);
        });
    }
};

export const removeExpenseAction = ({id} = {}) => {
    return {
        type: 'REMOVE_EXPENSE',
        id 
    }
};

export const editExpenseAction = ({id, amount, description, note, createdAt} = {}) => {
    return {
        type: 'EDIT_EXPENSE',
        payload: {
            id,
            amount,
            description,
            note,
            createdAt
        }
    }
};

export const setExpensesAction = (expenses = []) => {
    return {
        type: 'SET_EXPENSES',
        payload: expenses
    }
};

export const startSetExpensesAction = () => {
    const expenses = [];
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpensesAction(expenses));
        }).catch((error) => {
            console.log("Error retreiving expenses", error);
        });
    };
};