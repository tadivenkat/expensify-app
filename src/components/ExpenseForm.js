import React from 'react';
import moment from 'moment';
import DatePicker from 'react-datepicker';


class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            id: props.expense ? props.expense.id : '',
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? props.expense.amount / 100 : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt).valueOf() : moment().valueOf(),
            error: ''
        };
    };
    
    onDescriptionChange = (event) => {
        const description = event.target.value;
        this.setState((currentState) => {
            return {
                description
            };
        });
    };

    onNoteChange = (event) => {
        const note = event.target.value;
        this.setState((currentState) => {
            return {
                note
            };
        });
    };

    onAmountChange = (event) => {
        const amount = event.target.value;
        if (!amount || amount.match(/^\d*(\.\d{0,2})?$/)) {
            this.setState((currentState) => {
                return {
                    amount
                };
            });
        }
    }

    onDateChange = (date) => {
        this.setState((currentState) => {
            return {
                createdAt: moment(date).valueOf()
            };
        });
    }

    onFormSubmit = (event) => {
        event.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => {
               return {
                   error: 'Please enter description and amount' 
               };
            });
        } else {
            this.setState(() => {
                return {
                    error: '' 
                };
            });
            // Construct expense object and set it on props to make it accessible to the parent component
            this.props.onSubmit({
                id: this.state.id,
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                note: this.state.note,
                createdAt: this.state.createdAt
            });
        }
    }
    render() {
        return (
            <form className="form" onSubmit={this.onFormSubmit}>
                {this.state.error && <p className="bg-danger text-white">{this.state.error}</p>}
                <div className="form-group">
                    <label for="description">Description</label>
                    <input 
                        type="text"
                        id="description"
                        className="form-control"
                        placeholder="Enter Description" 
                        autoFocus 
                        value={this.state.description} 
                        onChange={this.onDescriptionChange} />
                </div>
                <div className="form-group">
                    <label for="amount">Amount</label>
                    <input
                        id="amount"
                        type="number"
                        className="form-control" 
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}/>
                </div>
                <div className="form-group">
                    <label for="note">Note</label>
                    <textarea 
                        id="note"
                        name="note"
                        className="form-control" 
                        placeholder="Add a Note for this Expense"
                        rows="7"
                        value={this.state.note}
                        onChange={this.onNoteChange}>
                    </textarea>
                </div>
                <div className="form-group">
                    <label for="createdAt">Created At</label>
                    <DatePicker
                        id="createdAt"
                        name="createdAt"
                        className="form-control" 
                        todayButton={"Today"}
                        selected={new Date(this.state.createdAt)} 
                        onChange={this.onDateChange}
                        title="Created At"
                        dateFormat="dd-MMM-yyyy"/>
                </div>
                <div>
                    <button onClick={this.onFormSubmit} className="btn btn-primary">Save Expense</button>
                </div>                    
            </form>                                        
        );
    }
}

export default ExpenseForm;