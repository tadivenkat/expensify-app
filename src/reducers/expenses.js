
export default (state = [], action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.payload];
        case 'REMOVE_EXPENSE':
            return state.filter(({id}) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.payload.id) {
                    return {...expense, ...action.payload};
                } else {
                    return expense;
                }
            });
        default: 
            return state;
    }
};