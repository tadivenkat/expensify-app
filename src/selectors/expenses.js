// Return subset of expenses applying the filters.

export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        let startDateMatch = false;
        let endDateMatch = false;
        const textMatch = expense.description.toLowerCase().indexOf(text.toLowerCase()) >= 0;
        if (!!startDate) {
            if (!!expense.createdAt) {
                startDateMatch = startDate <= expense.createdAt;
            } else {
                startDateMatch = true;
            }
        } else {
            startDateMatch = true;
        }
        if (!!endDate) {
            if (!!expense.createdAt) {
                endDateMatch = endDate >= expense.createdAt;
            } else {
                endDateMatch = true;
            }
        } else {
            endDateMatch = true;
        }
        return startDateMatch && endDateMatch && textMatch;
    }).sort((expense1, expense2) => {
        switch (sortBy) {
            case 'amount':
                return expense1.amount - expense2.amount;
            case 'date':
                return expense1.createdAt - expense2.createdAt;
            default:
                return expense1.amount - expense2.amount;
        }
    });
};