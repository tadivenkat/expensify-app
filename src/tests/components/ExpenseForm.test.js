import React from 'react';
import {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';
import ExpenseForm from '../../components/ExpenseForm';
import expenses from '../fixtures/expenses';

test('Testing ExpenseForm component snapshot', () => {
    const wrapper = shallow(<ExpenseForm />);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('Testing ExpenseForm component snapshot with default expense property', () => {
    const wrapper = shallow(<ExpenseForm expense={expenses[0]}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('Should render error message when the form is invalid', () => {
    const wrapper = shallow(<ExpenseForm/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('error').length).toBeGreaterThan(0);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('Should update description state when input changes', () => {
    const wrapper = shallow(<ExpenseForm/>);
    const value = "Test Expense Description";
    wrapper.find('input').at(0).simulate('change', { target: {value}});
    expect(wrapper.state('description')).toBe(value);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('Should update note state when textarea changes', () => {
    const wrapper = shallow(<ExpenseForm/>);
    const value = "Test Expense Note";
    wrapper.find('textarea').simulate('change', { target: {value}});
    expect(wrapper.state('note')).toBe(value);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('Should update amount state when the amount input changes', () => {
    const wrapper = shallow(<ExpenseForm/>);
    const value = "30";
    wrapper.find('input').at(1).simulate('change', { target: {value}});
    expect(wrapper.state('amount')).toBe(value);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('Should NOT update amount state when the amount input has invalid value', () => {
    const wrapper = shallow(<ExpenseForm/>);
    const value = "12.123";
    wrapper.find('input').at(1).simulate('change', { target: {value}});
    expect(wrapper.state('amount')).toBe('');
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('Testing form submit with valid data', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<ExpenseForm expense={expenses[0]} onSubmit={onSubmitSpy}/>);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    // No errors as the form is valid
    expect(wrapper.state('error')).toBe('');
    expect(onSubmitSpy).toHaveBeenLastCalledWith(expenses[0]);
});

test('Should update the date property of the state on change of date', () => {
    const newDate = 4123;
    const wrapper = shallow(<ExpenseForm/>);
    wrapper.find("DatePicker").prop("onChange")(newDate);
    expect(wrapper.state('createdAt')).toBe(newDate);
});