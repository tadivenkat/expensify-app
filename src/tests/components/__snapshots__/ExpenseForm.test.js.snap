// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should NOT update amount state when the amount input has invalid value 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      for="description"
    >
      Description
    </label>
    <input
      autoFocus={true}
      className="form-control"
      id="description"
      onChange={[Function]}
      placeholder="Enter Description"
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="amount"
    >
      Amount
    </label>
    <input
      className="form-control"
      id="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="number"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="note"
    >
      Note
    </label>
    <textarea
      className="form-control"
      id="note"
      name="note"
      onChange={[Function]}
      placeholder="Add a Note for this Expense"
      rows="7"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="createdAt"
    >
      Created At
    </label>
    <DatePicker
      allowSameDay={false}
      className="form-control"
      dateFormat="dd-MMM-yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      id="createdAt"
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      name="createdAt"
      nextMonthButtonLabel="Next month"
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      readOnly={false}
      renderDayContents={[Function]}
      selected={1970-01-01T00:00:00.000Z}
      shouldCloseOnSelect={true}
      showMonthYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      title="Created At"
      todayButton="Today"
      withPortal={false}
    />
  </div>
  <div>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`Should render error message when the form is invalid 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      for="description"
    >
      Description
    </label>
    <input
      autoFocus={true}
      className="form-control"
      id="description"
      onChange={[Function]}
      placeholder="Enter Description"
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="amount"
    >
      Amount
    </label>
    <input
      className="form-control"
      id="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="number"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="note"
    >
      Note
    </label>
    <textarea
      className="form-control"
      id="note"
      name="note"
      onChange={[Function]}
      placeholder="Add a Note for this Expense"
      rows="7"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="createdAt"
    >
      Created At
    </label>
    <DatePicker
      allowSameDay={false}
      className="form-control"
      dateFormat="dd-MMM-yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      id="createdAt"
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      name="createdAt"
      nextMonthButtonLabel="Next month"
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      readOnly={false}
      renderDayContents={[Function]}
      selected={1970-01-01T00:00:00.000Z}
      shouldCloseOnSelect={true}
      showMonthYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      title="Created At"
      todayButton="Today"
      withPortal={false}
    />
  </div>
  <div>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`Should render error message when the form is invalid 2`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="bg-danger text-white"
  >
    Please enter description and amount
  </p>
  <div
    className="form-group"
  >
    <label
      for="description"
    >
      Description
    </label>
    <input
      autoFocus={true}
      className="form-control"
      id="description"
      onChange={[Function]}
      placeholder="Enter Description"
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="amount"
    >
      Amount
    </label>
    <input
      className="form-control"
      id="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="number"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="note"
    >
      Note
    </label>
    <textarea
      className="form-control"
      id="note"
      name="note"
      onChange={[Function]}
      placeholder="Add a Note for this Expense"
      rows="7"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="createdAt"
    >
      Created At
    </label>
    <DatePicker
      allowSameDay={false}
      className="form-control"
      dateFormat="dd-MMM-yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      id="createdAt"
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      name="createdAt"
      nextMonthButtonLabel="Next month"
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      readOnly={false}
      renderDayContents={[Function]}
      selected={1970-01-01T00:00:00.000Z}
      shouldCloseOnSelect={true}
      showMonthYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      title="Created At"
      todayButton="Today"
      withPortal={false}
    />
  </div>
  <div>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`Should update amount state when the amount input changes 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      for="description"
    >
      Description
    </label>
    <input
      autoFocus={true}
      className="form-control"
      id="description"
      onChange={[Function]}
      placeholder="Enter Description"
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="amount"
    >
      Amount
    </label>
    <input
      className="form-control"
      id="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="number"
      value="30"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="note"
    >
      Note
    </label>
    <textarea
      className="form-control"
      id="note"
      name="note"
      onChange={[Function]}
      placeholder="Add a Note for this Expense"
      rows="7"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="createdAt"
    >
      Created At
    </label>
    <DatePicker
      allowSameDay={false}
      className="form-control"
      dateFormat="dd-MMM-yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      id="createdAt"
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      name="createdAt"
      nextMonthButtonLabel="Next month"
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      readOnly={false}
      renderDayContents={[Function]}
      selected={1970-01-01T00:00:00.000Z}
      shouldCloseOnSelect={true}
      showMonthYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      title="Created At"
      todayButton="Today"
      withPortal={false}
    />
  </div>
  <div>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`Should update description state when input changes 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      for="description"
    >
      Description
    </label>
    <input
      autoFocus={true}
      className="form-control"
      id="description"
      onChange={[Function]}
      placeholder="Enter Description"
      type="text"
      value="Test Expense Description"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="amount"
    >
      Amount
    </label>
    <input
      className="form-control"
      id="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="number"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="note"
    >
      Note
    </label>
    <textarea
      className="form-control"
      id="note"
      name="note"
      onChange={[Function]}
      placeholder="Add a Note for this Expense"
      rows="7"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="createdAt"
    >
      Created At
    </label>
    <DatePicker
      allowSameDay={false}
      className="form-control"
      dateFormat="dd-MMM-yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      id="createdAt"
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      name="createdAt"
      nextMonthButtonLabel="Next month"
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      readOnly={false}
      renderDayContents={[Function]}
      selected={1970-01-01T00:00:00.000Z}
      shouldCloseOnSelect={true}
      showMonthYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      title="Created At"
      todayButton="Today"
      withPortal={false}
    />
  </div>
  <div>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`Should update note state when textarea changes 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      for="description"
    >
      Description
    </label>
    <input
      autoFocus={true}
      className="form-control"
      id="description"
      onChange={[Function]}
      placeholder="Enter Description"
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="amount"
    >
      Amount
    </label>
    <input
      className="form-control"
      id="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="number"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="note"
    >
      Note
    </label>
    <textarea
      className="form-control"
      id="note"
      name="note"
      onChange={[Function]}
      placeholder="Add a Note for this Expense"
      rows="7"
      value="Test Expense Note"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="createdAt"
    >
      Created At
    </label>
    <DatePicker
      allowSameDay={false}
      className="form-control"
      dateFormat="dd-MMM-yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      id="createdAt"
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      name="createdAt"
      nextMonthButtonLabel="Next month"
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      readOnly={false}
      renderDayContents={[Function]}
      selected={1970-01-01T00:00:00.000Z}
      shouldCloseOnSelect={true}
      showMonthYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      title="Created At"
      todayButton="Today"
      withPortal={false}
    />
  </div>
  <div>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`Testing ExpenseForm component snapshot 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      for="description"
    >
      Description
    </label>
    <input
      autoFocus={true}
      className="form-control"
      id="description"
      onChange={[Function]}
      placeholder="Enter Description"
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="amount"
    >
      Amount
    </label>
    <input
      className="form-control"
      id="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="number"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="note"
    >
      Note
    </label>
    <textarea
      className="form-control"
      id="note"
      name="note"
      onChange={[Function]}
      placeholder="Add a Note for this Expense"
      rows="7"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="createdAt"
    >
      Created At
    </label>
    <DatePicker
      allowSameDay={false}
      className="form-control"
      dateFormat="dd-MMM-yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      id="createdAt"
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      name="createdAt"
      nextMonthButtonLabel="Next month"
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      readOnly={false}
      renderDayContents={[Function]}
      selected={1970-01-01T00:00:00.000Z}
      shouldCloseOnSelect={true}
      showMonthYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      title="Created At"
      todayButton="Today"
      withPortal={false}
    />
  </div>
  <div>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`Testing ExpenseForm component snapshot with default expense property 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      for="description"
    >
      Description
    </label>
    <input
      autoFocus={true}
      className="form-control"
      id="description"
      onChange={[Function]}
      placeholder="Enter Description"
      type="text"
      value="pen"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="amount"
    >
      Amount
    </label>
    <input
      className="form-control"
      id="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="number"
      value={100}
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="note"
    >
      Note
    </label>
    <textarea
      className="form-control"
      id="note"
      name="note"
      onChange={[Function]}
      placeholder="Add a Note for this Expense"
      rows="7"
      value="test note for test expense"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      for="createdAt"
    >
      Created At
    </label>
    <DatePicker
      allowSameDay={false}
      className="form-control"
      dateFormat="dd-MMM-yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      id="createdAt"
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      name="createdAt"
      nextMonthButtonLabel="Next month"
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      readOnly={false}
      renderDayContents={[Function]}
      selected={1970-01-01T00:00:08.000Z}
      shouldCloseOnSelect={true}
      showMonthYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      title="Created At"
      todayButton="Today"
      withPortal={false}
    />
  </div>
  <div>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Save Expense
    </button>
  </div>
</form>
`;
